### ISSUE TITLE
IME composition cursor moves incorrectly in SDL 2.26.5, 2.24.2 and 2.0.22

### ISSUE THREAD

**Issue by Robyt3** [libsdl-org/SDL#7616]
For example when using the Japanese IME and typing "watashi" as "わたし" the caret should initially be positioned after the character "し". While the composition is active, the cursor can be moved. For example moving it one position to the left and pressing backspace should remove the character "た" from the composition while keeping the composition active.

What's broke in 2.26.5, 2.24.2 and 2.0.22 is that the cursor can no longer be moved all the way to the start. When you move the cursor all the way to the left, the `SDL_TEXTEDITING` event instead reports `Event.edit.start == 3`. Pressing the right arrow key at this point correctly moves the cursor to the correct position again (to the right of "わ") though.

It works correctly in versions 2.0.20, 2.0.18, 2.0.16 and 2.0.8. Here the event reports `Event.edit.start == 0` when the candidate cursor is all the way on the left side.

I'm using Windows 10 64 bit.

**Comment by slouken** [comment]
Can you use git bisect to find out where this broke?

**Comment by Robyt3** [comment]
d14a12638309a3cba55bbb245df3347379b86f8f is the first bad commit
```
commit d14a12638309a3cba55bbb245df3347379b86f8f
Author: Zach Reedy <86971250+zreedy@users.noreply.github.com>
Date:   Fri Mar 11 17:45:17 2022 -0500

    IME Composition Truncation + SDL_IsTextInputShown + SDL_ClearComposition (#5398)

    * Fixes for IME Composition Truncation + Addition of SDL_ClearComposition, SDL_IsTextInputShown

    * Fixed: Documentation and code style issues raised during code review.

 include/SDL_events.h                    |  15 ++++
 include/SDL_hints.h                     |  11 +++
 include/SDL_keyboard.h                  |  18 +++++
 src/dynapi/SDL_dynapi_overrides.h       |   2 +
 src/dynapi/SDL_dynapi_procs.h           |   2 +
 src/events/SDL_keyboard.c               |  16 ++++
 src/video/SDL_sysvideo.h                |   2 +
 src/video/SDL_video.c                   |  18 +++++
 src/video/windows/SDL_windowskeyboard.c | 129 +++++++++++++++++++++++++++++---
 src/video/windows/SDL_windowskeyboard.h |   2 +
 src/video/windows/SDL_windowsvideo.c    |   2 +
 src/video/windows/SDL_windowsvideo.h    |   5 +-
 12 files changed, 210 insertions(+), 12 deletions(-)
```

**Comment by slouken** [comment]
@zreedy, can you take a look?

**Comment by zreedy** [comment]
I'm actually going to be looking into IME integration in our software again within the next week or so. I'll verify this issue and send a PR upstream.
Last time I tested Japanese IME this behavior seemed to work appropriately. Perhaps there's a byte-counting issue with kana vs kanji?

**Comment by slouken** [comment]
@zreedy, ping?

**Comment by wareya** [comment]
I ran into this with one of my projects (https://github.com/wareya/waui) and can replicate it on my system (windows 10, Google IME) but someone I know couldn't replicate it on theirs (OSX), so it's platform-specific. I haven't found anyone with Microsoft IME to test if it's a Google IME specific problem yet. I tried turning Microsoft IME on myself but couldn't figure out how to switch from direct input to japanese input mode.
